opencog/atoms/distvalue/DistributionalValue.h

```diff
@@ -0,0 +1,157 @@
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _OPENCOG_DISTRIBUTIONAL_VALUE_H
+#define _OPENCOG_DISTRIBUTIONAL_VALUE_H
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <limits>
+
+#include <opencog/util/Counter.h>
+#include <opencog/atoms/distvalue/ConditionalDV.h>
+
+/** \addtogroup grp_atomspace
+ *	@{
+ */
+
+namespace opencog
+{
+
+typedef double strength_t;
+typedef double confidence_t;
+typedef double count_t;
+
+class DistributionalValue;
+
+typedef std::vector<double> DVec;
+
+typedef std::vector<double> Interval;
+
+typedef std::vector<Interval> DVKey;
+
+typedef std::vector<DVKey> DVKeySeq;
+
+typedef Counter<DVKey, double> DVCounter;
+
+typedef std::shared_ptr<const DistributionalValue> DistributionalValuePtr;
+
+class AtomSpace;
+
+class DistributionalValue
+	: public Value
+{
+
+	friend class ConditionalDV;
+
+	DVCounter _value;
+
+	// Disallow assignment -- truth values are immutable!
+	DistributionalValue& operator=(const DistributionalValue& rhs) {
+		throw RuntimeException(TRACE_INFO, "Cannot modify truth values!");
+	}
+
+
+public:
+	static count_t DEFAULT_K;
+
+	DistributionalValue();
+	DistributionalValue(const DVCounter&);
+	DistributionalValue(double, double);
+
+	const DVCounter& value() const { return _value; }
+
+	static DistributionalValuePtr UniformDistributionalValue(const DVKey&, int);
+	static DistributionalValuePtr UniformDistributionalValue(const DVKeySeq&, int);
+	static DistributionalValuePtr TRUE_TV();
+	static DistributionalValuePtr FALSE_TV();
+	static DistributionalValuePtr DEFAULT_TV();
+	static DistributionalValuePtr createDV(double, double);
+	static DistributionalValuePtr createDV(const DVCounter&);
+
+	static double to_conf(int c);
+	static int to_count(double);
+
+	bool is_uniform() const;
+
+	std::vector<double> get_mode() const;
+	std::vector<double> get_mean() const;
+	std::vector<double> get_var() const;
+
+	double get_fstord_mean() const;
+	DVec middle_of_interval(const DVKey&) const;
+
+	double get_mode_for(double) const;
+	double get_mean_for(double) const;
+	double get_var_for(double) const;
+
+	DistributionalValuePtr add_evidence(const DVKey&) const;
+	DistributionalValuePtr merge(DistributionalValuePtr) const;
+	DistributionalValuePtr negate() const;
+
+	double min_count() const;
+	double max_count() const;
+
+	double total_count() const;
+	double get_confidence() const;
+
+	static double key_contained(const DVKey&, const DVKey&);
```

Comment:
@rTreutlein, I wonder about https://www.boost.org/doc/libs/1_69_0/libs/numeric/interval/doc/interval.htm ? Would that be overkill in your opinion? Or maybe at least using their terminology.

Reply:
I looked at that in the beginning and decided against it but maybe it's worth considering again. The main issue is that they don't support multidimensional Intervals or actually they might if I have an Interval over a vector of doubles that might work. I'll have to try that and see how it works out. But in any case using their terminology makes sense.

Reply:
Looking the git history of the project it seems rather mature. It's been in dev for 17 years and hasn't got a code change for at most 2 years.